libeos-update-server/server.c

```diff
@@ -0,0 +1,401 @@
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Philip Withnall <withnall@endlessm.com>
+ */
+
+#include <gio/gio.h>
+#include <glib.h>
+#include <glib-object.h>
+#include <libsoup/soup.h>
+
+#include <libeos-update-server/repo.h>
+#include <libeos-update-server/server.h>
+
+/**
+ * SECTION:server
+ * @title: HTTP server
+ * @short_description: HTTP server for multiple bare OSTree repositories
+ * @include: libeos-update-server/server.h
+ *
+ * A server that sits on top of zero or more bare repositories and lies to
+ * clients about the repositories’ mode, so it is possible to do pulls
+ * from this repository.
+ *
+ * Each repository is served under its #EusRepo:root-path prefix.
+ *
+ * Since: UNRELEASED
+ */
+
+/**
+ * EusServer:
+ *
+ * A server which handles serving zero or more #OstreeRepos at specified paths.
+ *
+ * Since: UNRELEASED
+ */
+struct _EusServer
+{
+  GObject parent_instance;
+
+  SoupServer *server;  /* owned */
+  GPtrArray *repos;  /* (element-type EusRepo), owned */
+
+  guint pending_requests;
+  gint64 last_request_time;
+};
+
+G_DEFINE_TYPE (EusServer, eus_server, G_TYPE_OBJECT)
+
+typedef enum
+{
+  PROP_SERVER = 1,
+  PROP_PENDING_REQUESTS,
+  PROP_LAST_REQUEST_TIME,
+} EusServerProperty;
+
+static GParamSpec *props[PROP_LAST_REQUEST_TIME + 1] = { NULL, };
+
+static void request_read_cb (SoupServer        *soup_server,
+                             SoupMessage       *message,
+                             SoupClientContext *client,
+                             gpointer           user_data);
+static void request_finished_cb (SoupServer        *soup_server,
+                                 SoupMessage       *message,
+                                 SoupClientContext *client,
+                                 gpointer           user_data);
+static void request_aborted_cb (SoupServer        *soup_server,
+                                SoupMessage       *message,
+                                SoupClientContext *client,
+                                gpointer           user_data);
+
+static void
+eus_server_init (EusServer *self)
+{
+  self->repos = g_ptr_array_new_with_free_func (g_object_unref);
+}
+
+static void
+eus_server_constructed (GObject *object)
+{
+  EusServer *self = EUS_SERVER (object);
+
+  G_OBJECT_CLASS (eus_server_parent_class)->constructed (object);
+
+  g_signal_connect (self->server, "request-read", (GCallback) request_read_cb, self);
```

Comment:
Good catch; we should check. We can’t create a `SoupServer` ourselves, since the caller is the one who needs to call `soup_server_listen()` on it.